Introduce PauliLindbladMap

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This is the second PR in a series of 3 introducing `QubitSparsePauli`, `QubitSparsePauliList`, and `PauliLindbladMap`. The first PR, which this PR depends on, is #14283 .

This introduces the data structure of the class `PauliLindladMap`, building on `QubitSparsePauliList`.

### Details and comments

This depends on #14283

DanPuzzuoli committed:
Introduce PauliLindbladMap

This is a roll-up of Dan's work of building the `PauliLindbladMap` on
top of [1].

[1]: 09e0de00bf: Add QubitSparsePauli and QubitSparsePauliList classes to quantum_info

Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>

jakelishman commented:
I rebased and rolled up this PR over the just-merged #14283, just to help speed-up the review process from my side - Dan/Sam, if you need to pull locally, your branches will be out of sync, so you'll need to make sure your branches are correctly at the previous tip commit 63477fdd1a70b2469eccf4b6bf7e164fcbaedfd2 (so we don't drop any changes you might have locally), then do `git fetch <dan's remote>` followed by `git reset --hard <dan's remote>/pauli-lindblad-map-v3` to forcibly reset your branches to this.

If you're worried, you can branch off locally first `git checkout -b old-pauli-lindblad-map pauli-lindblad-map-v3` before you start the destructive actions on the branch with `git reset --hard`, but as long as your branch was at the tip 63477fdd1a70b2469eccf4b6bf7e164fcbaedfd2, there shouldn't be problems.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -845,7 +907,7 @@ impl<'py> FromPyObject<'py> for Pauli {
 pub struct PyQubitSparsePauli {
-    inner: QubitSparsePauli,
+    pub inner: QubitSparsePauli,
 }
```
This is unsafe from a memory-safety perspective: you have `unsafe` blocks in `PyQubitSparsePauli` that rely on the inner object being immutable (or more specifically, its internal allocations not moving) within the lifetime of the outer object.

Specifically: the methods that return Numpy arrays that are backed by the Rust-space allocations can become memory-unsafe with this, if a user of `PyQubitSparsePauli` were to use the `pub` visibility on `inner` to move it and drop it while a Numpy array was still live.  The Python-space lifetimes only protect the existence of the outer `PyQubitSparsePauli`, and not its inner object.

What do you need the `pub` for here?  We can likely find ways that give you the access you need that remain safe.

---

As a follow-up: I realise that this is _way_ too easy to get wrong now, and you have to have huge amounts of knowledge in mind about the entirety of the `SparseObservable` data structures to keep this safe.  I'll follow up with a PR to `SparseObservable` and to all these new structures that modifies how the Numpy array-access sharing is done to make this much harder to get wrong, at the cost of a small runtime penalty that I expect will be close-enough to unmeasurable anyway.

DanPuzzuoli replied:
In the follow up PR I had instances of wanting to access the rust-space object `QubitSparsePauli` from within methods of `PyPauliLindbladMap`.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -438,6 +438,49 @@ impl QubitSparsePauliList {
 
+type RawParts = (Vec<Pauli>, Vec<u32>, Vec<usize>);
+
+pub fn raw_parts_from_sparse_list(
+    iter: Vec<(String, Vec<u32>)>,
+    num_qubits: u32,
+) -> Result<RawParts, LabelError> {
```
It's fine to split it out like this.  That said, if you prefer, you can provide a safe and no-cost way of achieving this from `QubitSparsePauli` itself, like so:
```rust
impl QubitSparsePauliList {
    pub fn take(self) -> (Vec<Pauli>, Vec<u32>, Vec<usize>) {
        (self.paulis, self.indices, self.boundaries)
    }
}
```
then you can have leave `QubitSparsePauliList::from_raw_parts` as is, and have the new callers of `raw_parts_from_sparse_list` be like
```rust
fn whatever(num_qubits: u32) -> Result<T, LabelError> {
    let (paulis, indices, boundaries) = QubitSparsePauliList::from_raw_parts(...)?.take();
}
```

Having `take` take `self` by ownership means that you can absolve yourself of coherence issues in the API after that - it's a way of providing safe API access to _completely_ destructure the `struct` into its constituent parts without having to give `pub` access to fields that can't be `pub` while the struct is still implying the semantics.

DanPuzzuoli replied:
The main reason I split this method out is that the same code is needed in `PauliLindbladMap`.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+};
+
+/// A Pauli Lindblad map that stores its data in a qubit-sparse format. Note that gamma,
+/// probabilities, and non_negative_rates is
```
Looks like something's got truncated here.

DanPuzzuoli replied:
Thanks - just completed the sentence.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+        let qubit_sparse_pauli_list: QubitSparsePauliList =
+            QubitSparsePauliList::new(num_qubits, paulis, indices, boundaries)?;
+        Self::new(rates, qubit_sparse_pauli_list)
+    }
```
The idea of `from_raw_parts` in `SparseObservable` is to be as close to a zero-cost constructor as possible.

This method is identical both in function and performance to `PauliLindbladMap::new(rates, QubitSparsePauliList::new(num_qubits, paulis, indices, boundaries)?)`, and it's (imo) more idiomatic rust not to hide that fact, but just have the caller spell it themselves, which it's not much more typing already.

DanPuzzuoli replied:
True - dropping this method and updating the one place it is used to use your suggested line.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            probabilities,
+            non_negative_rates,
+        }
+    }
```
Typically the purpose of `unsafe` constructors is to allow zero-computation methods of construction to callers who don't have privileged access to the internals of the struct.  This method still does some fairly heavy calculations (including some memory allocations), so unless it's _super_ necessary, it's possibly not worth the risk of adding an `unsafe` method.

I don't think you use this anywhere, so it might be cleanest just to remove it.

DanPuzzuoli replied:
This does get used in the followup PR, though could easily be replaced. It's true that all it does is not check if `rates.len() == qubit_sparse_pauli_list.len()`.

DanPuzzuoli replied:
Okay, deleted this.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            .map(|(i, rate)| GeneratorTermView {
+                rate: *rate,
+                qubit_sparse_pauli: self.qubit_sparse_pauli_list.term(i),
+            })
```
This is totally fine as-is, but just to comment on alternatives: Rust has a `zip` method similar to Python's, so this is doable as
```rust
self.rates.iter().zip(self.qubit_parse_pauli_list.iter()).map(|(&rate, qubit_sparse_pauli)| {
    GeneratorTermView { rate, qubit_sparse_pauli }
})
```
(or something like that) too.  No need to change it, just saying it's an option for the future if you prefer that form / if you have a struct where you don't have constant-time random access (which `QubitSparsePauliList::term` implies).

DanPuzzuoli replied:
Nice ty - I copied in your suggestion. I need to learn how to use rust idiomatically :). I started using zip at some point but may have put this in before that.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+/// See [PyPauliLindbladMap] for detailed docs.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PauliLindbladMap {
+    /// The rates of each abstract term in the generator sum.  This has as many elements as
+    /// terms in the sum.
+    rates: Vec<f64>,
+    /// A list of qubit sparse Paulis corresponding to the rates
+    qubit_sparse_pauli_list: QubitSparsePauliList,
+    /// The gamma parameter.
+    gamma: f64,
+    /// Probability of application of a given Pauli operator in product form. Note that if the
+    /// corresponding rate is less than 0, this is a quasi-probability.
+    probabilities: Vec<f64>,
+    /// List of boolean values for the statement rate >= 0 for each rate in rates.
+    non_negative_rates: Vec<bool>,
```
Some of the choices of stored calculated values are slightly strange to me here, but I don't have the domain knowledge to say if it's _truly_ worthwhile - you're better placed to determine that than me.

Given that `gamma`, `probabilities` and `non_negative_rates` all have a relatively simple derivation from `rates`, is it completely necessary to store them in the class, or could they have methods that calculate them, and the caller stores them?  There's a few considerations I'm thinking of:
- storing them in the class means your data-at-rest form is always paying the cost of storing them
- the derived quantities _must_ be updated whenever the class it mutated, even if you might be applying a sequence of mutations, and the derivation can be done more simply/more performantly if it's done in a single step right at the end
- you need the derived quantities `gamma` and `probabilities` frequently to persist between calls

I _suspect_ that your concern is about the latter point, in particular w.r.t the `sample` method that's about to be introduced.  I can't speak for the expected usage patterns that would motivate one way or the other for the previous two points, but assuming that you _do_ pay a cost for them, and `sample` (and friends) is the main reason to have this necessary cache constantly, then there is a way around it:
- _don't_ store the derived quantities
- introduce a new struct like `PauliLindbladMapSampler` that calculates the derived quantities on construction, and stores them only during its lifetime, which lets the caller manage how long they ought to last for, and persist them
- Give the new struct the methods that depend on the derived quantities

In pure Rust space, that can often give you the best of both worlds - you don't pay for memory usage until the point you use it, you don't pay extra for intermediate mutations to the base object, the caller has the power to define exactly how long they need the derived quantities to live for, and you still don't have to re-derive the quantities on _every_ call to `sample`.

The trade-off is that it's a bit more implementation effort on this side, and it can be a bit more complex when you're exposing it to Python - it's still very doable, but there's more typing needed to do it.

---

The upshot of this is: if you're pretty sure that an instance of this `struct` _always_ terminates in a bunch of calls that necessarily use the derived quantities a bunch, and the cost of updating the derived quantities under mutation is no more expensive than deriving them only once at the point of use (or you frequently sample a partially constructed object, then add more generators, then sample more, etc), then there's no real harm in keeping the derived quantities here in the struct.  If you're not sure about that, it's safer from an expansion perspective, and potentially gives more power to the caller to _not_ store the derived quantities, but expose an API that lets the caller explicitly say when and for how long the derived quantities are actually needed.

DanPuzzuoli replied:
So I do think that the main use cases we care about fit into the "terminates in a bunch of calls that necessarily use the derived quantities a bunch", as the main performance concern is that of sampling. While the subsequent PR does introduce some functionality beyond sampling, we already primarily think of this class as "the thing that does sampling", so it feels like it would be redundant to have an additional class made purely for drawing samples.

Some use cases to think about:
1. For PEC, one computes the inverse map once, and then it gets sampled from many times.
2. For a protocol like error amplification, the noise needs to be re-scaled (rates multiplied with some constant) with various scaling values. One issue here is that the `scale_rates` method in the subsequent PR returns a totally new object (with a copy of `QubitSparsePauliList`). So, you have a situation where sampling from a rate-scaled version requires fully copying the whole operator data structure.
3. There are additional more complex versions of case 2 that lead to having the `scale_factor` be a vector of the same length as `rates`, and "scaling" is then entry-wise multiplication.

All of the above cases terminate in a bunch of sampling, but cases 2 and 3 suffer from the issue of possibly copying the internal `QubitSparsePauliList` for every scale factor (or implementing Ian's suggestion, which requires recomputing the derived quantities for each `sample` call anyway). One option to get around this that Ian suggested (possibly in a comment to the subsequent PR by now) is to have an optional `scale_factor` argument to `sample` that will compute the scaled the rates, re-compute the derived quantities, and use those in sampling (the underlying rates in the object however, are unchanged).

I guess an alternative option to get around the `QubitSparsePauliList` copying issue in cases 2 & 3 would be if everything is immutable? Then they could just be passed around/shared by reference? I think I had thought of doing this at one point but some of the methods for building the objects used `add_term` which requires mutability.

jakelishman replied:
For the future: we intend to split these out in a future version of `PauliLindbladMap`, to make the objects lighter to manipulate, especially want sampling under different rates/masks.  See #14528.

They're staying in the object for this PR, because it still _works_, and we're just not exposing them publicly to Python space in the interface.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            self.qubit_sparse_pauli_list
+                .add_qubit_sparse_pauli(new_pauli.view())?;
+            Ok(())
+        }
```
It's generally best to keep `unsafe` blocks scoped _exactly_ as small as they can be.  The trick that perhaps you missed here is that `unsafe { }` introduces a scope, but it still evaluates to an expression, so you want
```rust
// SAFETY: ...
let new_pauli = unsafe { QubitSparsePauli::new_unchecked(...) };
```
(idiomatically, you introduce the comment about the safety of a use of `unsafe` with `SAFETY:`)

You can't put the `let` inside the `unsafe`, or its lifetime ends when the closing brace happens.

DanPuzzuoli replied:
Nice - thanks! Changed it.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+    let p: f64 = w / g;
+    let nnr: bool = rate >= 0.0;
+    (g, p, nnr)
+}
```
Very minor: `f64` is `Copy` (can be freely bitwise copied in memory), so it's a bit unusual to define a function that takes `&f64`.  Usually you just take `f64`, and then when you need to do things like `rates.iter().map(derived_values_from_rate)` or the like, you can instead to `rates.iter().copied().map(derived_values_from_rate)`, etc.

DanPuzzuoli replied:
Thanks, changed.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+        non_negative_rates.push(nnr);
+    }
+    (gamma, probabilities, non_negative_rates)
+}
```
Accepting `&Vec<T>` is very rarely the best type to accept, idiomatically.  If nothing else, you typically want `rates: &[f64]` which (if there's no inlining) is technically one pointer more direct than `&Vec<T>`.  In some cases, you can even accept `rates: impl IntoIterator<Item=f64>` and have it generic, but there's no need to introduce generics here.

Second point: you probably want `Vec::with_capacity(rates.len())` for both `probabilities` and `non_negative_rates`.

DanPuzzuoli replied:
Implemented.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+/// A view object onto a single generator term of a `PauliLindbladMap`.
+///
+/// The lengths of `paulis` and `indices` are guaranteed to be created equal, but might be zero
+/// (in the case that the term is proportional to the identity).
```
Your comment is out of date here.  I think there's other places in `impl` methods in the block above too, but I only quickly scanned them.

DanPuzzuoli replied:
Deleted... didn't see others.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+    /// The real rate of the term.
+    rate: f64,
+    qubit_sparse_pauli: QubitSparsePauli,
+}
```
This has Sphinx-style documentation, but isn't a `pyclass`.  Is that intentional?

DanPuzzuoli replied:
No, just habit. I just removed the `:class:` from both lines.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            qubit_sparse_pauli,
+        }
+    }
+
```
As far as I can tell, there's no way that `rate` and `qubit_sparse_pauli` can ever become incoherent relative to each other, and there's no validity requirements other than the base types of the objects.

If that's guaranteed to always be the case and this is always a very simple record type (I suspect it is?), then this is a case where you can safely make the fields `rate` and `qubit_sparse_pauli` public, and avoid defining a separate `new` constructor.

DanPuzzuoli replied:
Yes true. It's only currently used int his file so no need to even make the fields public. Deleted the `new` method.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            vec![0, self.paulis().len()],
+        )?;
+        PauliLindbladMap::new(vec![self.rate], qubit_sparse_pauli_list)
+    }
```
Since you've already got a `QubitSparsePauli` in the inner object, there should be an infallible way to lift that into a `QubitSparsePauliList`, I believe?  There's nothing that `List` adds that make a single term represented as a list suddenly incoherent, and similarly nothing about the pairing with a single rate that can either.

Assuming I'm right about that, it's nicer to reflect that in the typing, and have this method return `PauliLindbladMap` infallibly; your API should already be guaranteeing that if you hold an object of type `QubitSparsePauli`, then it's internally valid, so you don't need to check again in these methods.

The idea with defining various partial methods in Rust, like why `SparseObservable` has several constructors with different error returns, why various lifting methods explicit exist, etc, is to let us define APIs that properly represent exactly which operations are fallible and which aren't, so we can statically and compiler-enforced avoid coherence checks on higher-order object construction.

DanPuzzuoli replied:
You're right, a method already existed for this. Changed it to:
```
PauliLindbladMap::new(vec![self.rate], self.qubit_sparse_pauli.to_qubit_sparse_pauli_list())
```

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+
+    fn to_label(&self) -> PyResult<String> {
+        Ok(self.inner.view().to_sparse_str())
+    }
```
Same question about publicly exposing this method as I had with `QubitSparsePauli`: is it necessary (because of its linking of the term `label` to a different mostly-internal format)?

DanPuzzuoli replied:
Yep, this method isn't used by anything and isn't necessary. Deleted!

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            .map(|bit| bit.py_label())
+            .collect();
+        PyString::new(py, string.as_str())
+    }
```
Similar question about these guys as with `QubitSparsePauli` (where you deleted it).  I imagine you've got more desire for it here?  If so: it'd be nice to include an example in the documentation illustrating what you mean on lines 506-507.

DanPuzzuoli replied:
Actually I also think this is unnecessary. Deleted :P .

Edit: I had originally deleted this but decided to put it back. I've added an example.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+            .inner
+            .read()
+            .map_err(|_| InnerReadError)?;
+        let inner = PauliLindbladMap::new(rates.clone(), qubit_sparse_pauli_list.clone())?;
```
You probably don't need to clone the `rates` here.  It's not super ideal that we have to clone the `QubitSparsePauliList` on input, but not the end of the world right now - can always be addressed later.

DanPuzzuoli replied:
Okay dropping the cloning of the rates. Yeah I don't like cloning the operator list either but I wasn't sure how to get around that. Is it correct that if we demand it is passed as a mutable reference we could guarantee no one else has access to change it anymore? (What would be idiomatic here?)

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+    fn get_qubit_sparse_pauli_list(&self) -> PyQubitSparsePauliList {
+        let inner = self.inner.read().unwrap();
+        inner.qubit_sparse_pauli_list.clone().into()
+    }
```
This is almost certainly not what you're after for performance: the `qubit_sparse_pauli_list` itself is cloned on each attribute access; it's like you're doing a Python-space `deepcopy` on the underlying list each time you access the attribute.  This is going to make `plm.qubit_sparse_pauli_list` attribute accesses _really_ expensive in Python space.

It's not immediately clear to me what your best bet is going to be here - you can't just expose internals of Rust objects to Python space safely without having designed in the capability from a fairly early point.

If you can get away with it for the first release, you're _probably_ best making this a regular method (so not a `getter`), giving it a slightly different name to make it clear it makes a copy, and relying on callers to eat the cost of the copy if they need the internal access.  We can potentially work out some new method of structuring the data types later to make it easier to make a raw getter in the future.

In particular: it would be easier to imagine such a situation if the derived quantities from `rates` _aren't_ stored, but are calculated only at the point of use.  That's not _necessary_, but my first thoughts for improving the attribute access here want it.  We can find a way around that if we need to.

DanPuzzuoli replied:
Okay sounds good. I've dropped the `getter` decorator and renamed the method to:
```
get_qubit_sparse_pauli_list_copy
```

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+///
+/// .. math::
+///
+///     \Lambda = \prod_{P \in K}\exp\left(\lambda_P(P \cdot P - \cdot)\right).
```
Minor style: here and everywhere in the docs, I find the `\cdot` very hard to read; to me, it's a multiplication marker.  Perhaps `\circ` is a better choice? Compare this one using `\cdot`:

$$
\Lambda = \prod_{P \in K}\exp\left(\lambda_P(P \cdot P - \cdot)\right)
$$

to this one using `\circ`:

$$
\Lambda\bigl[\circ\bigr] = \prod_{P \in K}\exp\left(\lambda_P(P \circ P - \circ)\right)
$$

DanPuzzuoli replied:
I don't feel strongly about this personally. I naturally prefer `\cdot` as I've always used `\circ` to refer to composition, but I also agree that in this case the `\cdot` is weird looking appearing by itself after a minus sign. I'll change it to `\circ`. I've also changed the $\Lambda \rightarrow \Lambda\bigl[\circ\bigr]$.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/pauli_lindblad_map_class.rs:
```diff
@@ -0,0 +1,1177 @@
+        let out = unsafe { PyArray1::borrow_from_array(&arr, slf_.into_any()) };
+        out.readwrite().make_nonwriteable();
+        out
+    }
```
The safety guarantees on these really don't hold in the same way that they did in the smaller scale objects.  I'm a bit worried about this.

DanPuzzuoli replied:
Thoughts on how to resolve this?

DanPuzzuoli replied:
I ended up dropping the `#[getter]` decorator so that this no longer looks like an attribute. I've done the same with `get_gamma`.

jakelishman reviewed (commented):
Thanks for this Dan - I've left some comments on this here.  I can't comment too much on the exact maths and whatnot of the object without spending a fair amount more time thinking about it, so I'm largely trusting that the three (four?) of you have got it right for what you wanted.

My biggest concerns here are that the way the Rust/Python split works with with compound object of `PauliLindbladMap` and the inner `QubitSparsePauliList` aren't right - there have been problems introduced as this moved beyond the structures that `SparseObservable` had laid the groundwork for.  I've left the comments here for first discussion, and I'll think about more asynchronously about some of the points I raised here.

DanPuzzuoli committed:
making suggested changes

DanPuzzuoli committed:
formatting and re-adding pauli_labels method

DanPuzzuoli committed:
formatting

DanPuzzuoli commented:
@jakelishman I believe I've addressed all of the comments except for the final one about concern with the `get_probabilities` method. Any thoughts on what to do with this?

I've also added the method `get_inner` to `PyQubitSparsePauli` to return the inner, and no longer directly access it (it is no longer public).

DanPuzzuoli committed:
fixing forgotten inner issue

jakelishman committed:
Fix minor documentation copy/paste

jakelishman committed:
Make `gamma()` and `probabilities()` regular computing methods

This defines the interface of these functions to be regulard methods
that compute quantities.  This includes having `probabilities` return a
copy of a Numpy array, which is what will be needed in follow-ups
anyway; we don't want the Python interface to leak that certain
quantities are actually currently cached (since that will change in the
future).

A couple of Rust-space methods are made more "Rust-y" as a side effect.

jakelishman committed:
Fix clippy

jakelishman reviewed (approved):
Ok, I think this is all safe for the interface, in preparation for more completely moving to #14528 in Qiskit 2.2.
